### PR TITLE
Database configuration refactor

### PR DESCRIPTION
**Pull request by sagepe**
This PR adds variables for replica databases to the database configuration generated by the deployment scripts, derived from the `replica` value in the database JSON and some logic to work out the appropriate port.

It also includes a wider refactor that moves the generation of the various different formats out of the `deploy-vhost` script and into the `write_settings_file` function in `Deploy.pm` which feels like a more natural place as well as removing checks for whether we are on an internal network and an include of the older machine settings file used for a single associated variable.

### CODE REVIEW

File: bin/deploy-vhost [hunk 1]
```diff
@@ -135,16 +135,30 @@ sub get_pgpw_string {
 
 sub one_db_conf_php {
     my ($params, $database, $internal) = @_;
-    my $port_line = '';
+    my $port_line       = '';
```
Comment (dracos) [inline bin/deploy-vhost]: This line didn't change ;) `not recommended` is our coding standard on this, but only if I make another comment about something else actual to change :)

File: bin/deploy-vhost [hunk 2]
```diff
@@ -135,16 +135,30 @@ sub get_pgpw_string {
...
+    my $db_ro_host_line = '';
+    my $port_ro_line    = '';
     if ($params->{type} eq 'psql') {
         $port_line = "define('OPTION_$params->{prefix}_DB_PORT', $params->{port});";
+        if ($params->{replica}) {
+            $db_ro_host_line = "define('OPTION_$params->{prefix}_DB_RO_HOST', '$params->{replica}.ukcod.org.uk');";
+            # Hot standby usually on 5433, unless it's been failed over at some point.
+            if ($params->{port} eq '5432') {
+                $port_ro_line = "define('OPTION_$params->{prefix}_DB_RO_PORT', 5433);";
```
Comment (dracos) [inline bin/deploy-vhost]: To avoid the repetition of most of this line, set `$ro_port = $params->{port} == 5432 ? 5433 : 5432;` first and then use that? Happens again below as well.

File: bin/deploy-vhost [hunk 3]
```diff
@@ -135,16 +135,30 @@ sub get_pgpw_string {
...
 sub one_db_conf_php {
     my ($params, $database, $internal) = @_;
-    my $port_line = '';
+    my $port_line       = '';
+    my $db_ro_host_line = '';
+    my $port_ro_line    = '';
     if ($params->{type} eq 'psql') {
         $port_line = "define('OPTION_$params->{prefix}_DB_PORT', $params->{port});";
+        if ($params->{replica}) {
+            $db_ro_host_line = "define('OPTION_$params->{prefix}_DB_RO_HOST', '$params->{replica}.ukcod.org.uk');";
```
Comment (dracos) [inline bin/deploy-vhost]: For consistency (though I realise it's the same output), should this be `get_db_host($params->{replica})`?

File: bin/deploy-vhost [hunk 4]
```diff
@@ -133,140 +133,56 @@ sub get_pgpw_string {
...
+        host     => get_db_host($params, 0),
+        name     => $database,
+        username => $database,
+        password => "\${\\(pgpw('$pgpw_string'))}",
+    };
+
+    if ($params->{port}) {
+        $db_config->{port} = $params->{port};
+        if ($params->{replica}) {
+            $db_config->{replica_host} = "$params->{replica}.ukcod.org.uk";
```
Comment (dracos) [inline bin/deploy-vhost]: Ditto other comments around here, e.g. calling get_db_host.

File: bin/deploy-vhost [hunk 5]
```diff
@@ -133,140 +133,56 @@ sub get_pgpw_string {
...
+# further data as required. Store this in $conf->{database_configs}
 sub db_conf {
     my ($params, $database) = @_;
     if ($params->{type} ne 'mysql' && $params->{type} ne 'psql' && $params->{type} ne 'mongo'){
         die "unknown database type '$params->{type}'";
     }
 
-    # Strings for PHP configuration
-    $conf->{database_configs}{default} .= one_db_conf_php($params, $database, 1);
-    $conf->{database_configs}{external} .= one_db_conf_php($params, $database, 0);
```
Comment (dracos) [inline bin/deploy-vhost]: Doesn't really matter, but the removal of the external bits seems to me separate from the refactoring out of the code (indeed, you've done one last bit in another commit), it'd be clearer if they had been removed first on their own, then the code moved, iyswim.

File: bin/deploy-vhost [hunk 6]
```diff
@@ -133,140 +133,56 @@ sub get_pgpw_string {
...
+        $db_config->{port} = $params->{port};
+        if ($params->{replica}) {
+            $db_config->{replica_host} = "$params->{replica}.ukcod.org.uk";
+            if ($params->{port} eq '5432') {
+                $db_config->{replica_port} = 5433;
+            }
+            elsif ($params->{port} eq '5433') {
+                $db_config->{replica_port} = 5432;
+            }
+            elsif ($params->{port} eq '3306') {
```
Comment (dracos) [inline bin/deploy-vhost]: If you specify replica for a mysql, but don't specify the default port (as you haven't had to do that in the past for mysql), this won't work; should the mysql `elsif` a bit down do a check for replica, just to cover all the bases? Or just make port mandatory then don't need that elsif...

File: bin/deploy-vhost (hunk 1 above)
Comment (dracos) [reply bin/deploy-vhost]: It's removed by the 2nd commit anyway.

File: bin/deploy-vhost (hunk 3 above)
Comment (dracos) [reply bin/deploy-vhost]: Is then removed by 2nd commit.

File: bin/deploy-vhost [hunk 7]
```diff
@@ -133,112 +118,55 @@ sub get_pgpw_string {
...
+        username => $database,
+        password => "\${\\(pgpw('$pgpw_string'))}",
+    };
+
+    if ($params->{port}) {
+        $db_config->{port} = $params->{port};
+        if ($params->{replica}) {
+            $db_config->{replica_host} = "$params->{replica}.ukcod.org.uk";
+            if ($params->{port} eq '5432') {
+                $db_config->{replica_port} = 5433;
```
Comment (dracos) [inline bin/deploy-vhost]: Feels like this list of port mappings should be a hash, or a function in Deploy.pm that can be called here.

**Comment by sagepe** [comment]
Thanks for the comments - I've shifted the commits around into what feels like a better order: (1) remove external config, (2) refactor config generation, (3) add new replica data.

I've also replaced the port mapping with a function in `Deploy.pm` and removed the dependency on setting port in the JSON for configuring a replica - if no port is set and the database is MySQL it will assume 3306 and generate the replica's port based on this. (This seems like a safe assumption for now as anyone using MySQL without specifying a port will definiteley be on 3306 anyway.)
